src/DagData.cpp

```diff
@@ -0,0 +1,138 @@
+            for(uint32_t childIndex: dependenciesLookup.For(dag_index))
+            {
+                if (shouldProcess == nullptr || (*shouldProcess)(dag_index, childIndex))
+                    BufferAppendOne(&node_stack, heap, childIndex);
+            }
+        }
+    }
+
+    HeapFree(heap, node_visited_bits);
+    BufferDestroy(&node_stack, heap);
+}
+
+void FindDependentNodesFromRootIndices(MemAllocHeap* heap, const Frozen::Dag* dag, const Frozen::DagDerived* dagDerived, std::function<bool(int,int)>* shouldProcess, int32_t* searchRootIndices, int32_t searchRootCount, Buffer<int32_t>& results)
+{
+    struct LookupDependenciesFromFrozen
+    {
+        const Frozen::DagDerived* dagDerived;
+        const FrozenArray<int32_t>& For(int i) { return dagDerived->m_Dependencies[i];}
+    } dependencyLookup;
+
+    dependencyLookup.dagDerived = dagDerived;
+
+    FindDependentNodesFromRootIndices_Shared(heap,dag,dependencyLookup, shouldProcess, searchRootIndices, searchRootCount, results);
+    return;
+}
+
+void FindDependentNodesFromRootIndices(MemAllocHeap* heap, const Frozen::Dag* dag, Buffer<int32_t>* dependencyBuffers, std::function<bool(int,int)>* shouldProcess, int32_t* searchRootIndices, int32_t searchRootCount, Buffer<int32_t>& results)
+{
+    struct LookupDependenciesFromBuffers
+    {
+        Buffer<int32_t> *buffers;
+        const Buffer<int32_t>& For(int i) { return buffers[i];}
+    } dependencyLookup;
+
+    dependencyLookup.buffers = dependencyBuffers;
+    FindDependentNodesFromRootIndices_Shared(heap,dag,dependencyLookup, shouldProcess, searchRootIndices, searchRootCount, results);
+    return;
+}
+
+void FindAllOutputFiles(const Frozen::Dag* dag, HashSet<kFlagPathStrings>& outputFiles)
+{
+    int node_count = dag->m_NodeCount;
+    for (int32_t i = 0; i < node_count; ++i)
+        for (auto& outputFile : dag->m_DagNodes[i].m_OutputFiles)
+            HashSetInsert(&outputFiles, outputFile.m_FilenameHash, outputFile.m_Filename.Get());
+}
+
+void DagRuntimeDataInit(DagRuntimeData* data, const Frozen::Dag* dag, MemAllocHeap *heap)
+{
+    HashTableInit(&data->m_OutputsToDagNodes, heap);
+    HashTableInit(&data->m_OutputDirectoriesToDagNodes, heap);
+    for (int i = 0; i<dag->m_NodeCount; i++)
+    {
+        const Frozen::DagNode* node = dag->m_DagNodes + i;
+        for(auto &output: node->m_OutputFiles)
+            HashTableInsert(&data->m_OutputsToDagNodes, output.m_FilenameHash, output.m_Filename.Get(), i);
+        for(auto &output: node->m_OutputDirectories)
+            HashTableInsert(&data->m_OutputDirectoriesToDagNodes, output.m_FilenameHash, output.m_Filename.Get(), i);
+    }
+
+    // We currently don't populate m_OutputDirectories for all nodes.
+    // Some output directries still only show up in m_DirectoriesCausingImplicitDependencies.
+    // For those, we don't know which node they came from, so we ise the special index value of -1.
```

Comment:
nitpick: s/ise/use/

Reply:
Yes, a nitpick, but fixing it reminded me that I had some other unsaved cleanups to this file I had not yet pushed :) (`FindAllOutputFiles` above is unused and should be removed). Fixed.